src/sentry/debug_files/artifact_bundles.py

```diff
@@ -34,14 +50,374 @@
+
+        for bundle in bundles:
+            serialized_bundles.append(
+                {"id": bundle["id"], "timestamp": datetime.isoformat(bundle["timestamp"])}
+            )
+
+        return serialized_bundles
+
+
+T = TypeVar("T")
+
+
+class FlatFileIndex:
+    def __init__(self, store: FlatFileIndexStore):
+        self.store = store
+
+        # By default, a flat file index is empty.
+        self._bundles: Bundles = []
+        self._files_by_url: FilesByUrl = {}
+        self._files_by_debug_id: FilesByDebugID = {}
+
+    @staticmethod
+    def build(store: FlatFileIndexStore) -> Optional[FlatFileIndex]:
+        index = FlatFileIndex(store)
+
+        try:
+            index._build()
+            return index
+        except Exception as e:
+            sentry_sdk.capture_exception(e)
+            metrics.incr("artifact_bundle_flat_file_indexing.error_when_building")
+            return None
+
+    def _build(self) -> None:
+        loaded_index = self.store.load()
+
+        # If we don't find an index, we will silently not mutate the state, since it's initialized as empty.
+        if loaded_index is not None:
+            self._bundles = cast(Bundles, loaded_index.get("bundles", []))
+            self._files_by_url = cast(FilesByUrl, loaded_index.get("files_by_url", {}))
+            self._files_by_debug_id = cast(
+                FilesByDebugID, loaded_index.get("files_by_debug_id", {})
+            )
+
+    def merge(self, artifact_bundle: ArtifactBundle, archive: ArtifactBundleArchive) -> bool:
+        try:
+            # We want to index by url if the store has an identifier which is being indexed by release.
+            self._merge(artifact_bundle, archive, self.store.identifier.is_indexing_by_release())
+            return True
+        except Exception as e:
+            sentry_sdk.capture_exception(e)
+            metrics.incr("artifact_bundle_flat_file_indexing.error_when_merging")
+            return False
+
+    def _merge(
+        self, artifact_bundle: ArtifactBundle, archive: ArtifactBundleArchive, index_urls: bool
+    ):
+        bundle_index = self._update_bundles(artifact_bundle)
+
+        if index_urls:
+            self._iterate_over_urls(bundle_index, archive)
+        else:
+            self._iterate_over_debug_ids(bundle_index, archive)
+
+    def _update_bundles(self, artifact_bundle: ArtifactBundle) -> int:
+        if artifact_bundle.date_last_modified is None:
+            raise Exception("Can't index a bundle with no date last modified")
+
+        bundle_meta = BundleMeta(
+            id=artifact_bundle.id, timestamp=artifact_bundle.date_last_modified
+        )
+
+        bundle_index = self._index_of_bundle(artifact_bundle.id)
+        if bundle_index is None:
+            self._bundles.append(bundle_meta)
+            return len(self._bundles) - 1
+        else:
+            self._bundles[bundle_index] = bundle_meta
+            return bundle_index
```

Comment:
My draft of this early-returned on this edge-case.

Now you are inserting the index duplicated, however it seems that duplicates are still being cleaned up in `_update_bundle_references`.

Though I agree that this is a case that needs more thought related to having split debug-id / url indexes, or both in one file.

Reply:
The reason for such code is to cover the edge case in which an existing bundle is inserted with a new date last modified. In this case, we need to re-sort entries to maintain the "upload intent" order.

As a minor optimization we can check if the dates are equal, if so, we don't need to do anything.

Reply:
Oh okay, that is a good point.
It should be fairly simple to support the case where you want to "re-index, but really just bump the timestamp".

Reply:
We will only need to handle de-duplication on collection insertion and that's it.